debian/mongofiles.1
.\" Man page generated from reStructuredText.
.
.TH "MONGOFILES" "1" "Jun 21, 2018" "4.0" "mongerdb-manual"
.SH NAME
mongerfiles \- MongerDB GridFS Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Synopsis\fP
.IP \(bu 2
\fI\%Required Access\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Commands\fP
.IP \(bu 2
\fI\%Examples\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.IP "Mac OSX Sierra and Go 1.6 Incompatibility"
.sp
Users running on Mac OSX Sierra require the 3.2.10 or newer version
of  mongerfiles\&.
.UNINDENT
.UNINDENT
.SH SYNOPSIS
.sp
The \fI\%mongerfiles\fP utility makes it possible to manipulate files
stored in your MongerDB instance in GridFS objects from the
command line. It is particularly useful as it provides an interface
between objects stored in your file system and GridFS.
.sp
All \fI\%mongerfiles\fP commands have the following form:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongerfiles <options> <commands> <filename>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The components of the \fI\%mongerfiles\fP command are:
.INDENT 0.0
.IP 1. 3
\fI\%Options\fP\&. You may use one or more of
these options to control the behavior of \fI\%mongerfiles\fP\&.
.IP 2. 3
\fI\%Commands\fP\&. Use one of these commands to
determine the action of \fI\%mongerfiles\fP\&.
.IP 3. 3
A filename which is either: the name of a file on your local’s file
system, or a GridFS object.
.UNINDENT
.sp
Run \fI\%mongerfiles\fP from the system command line, not the \fBmonger\fP shell.
.sp
\fBIMPORTANT:\fP
.INDENT 0.0
.INDENT 3.5
For replica sets,
\fI\%mongerfiles\fP can only read from the set’s
primary\&.
.UNINDENT
.UNINDENT
.SH REQUIRED ACCESS
.sp
In order to connect to a \fBmongerd\fP that enforces authorization
with the \fB\-\-auth\fP option, you must use the
\fI\%\-\-username\fP and \fI\%\-\-password\fP options. The connecting user must possess, at a
minimum:
.INDENT 0.0
.IP \(bu 2
the \fBread\fP role for the accessed database when using the
\fBlist\fP, \fBsearch\fP or \fBget\fP commands,
.IP \(bu 2
the \fBreadWrite\fP role for the accessed database when using
the \fBput\fP or \fBdelete\fP commands.
.UNINDENT
.SH OPTIONS
.sp
Changed in version 3.0.0: \fI\%mongerfiles\fP removed the \fB\-\-dbpath\fP as well as related
\fB\-\-directoryperdb\fP and \fB\-\-journal\fP options. To use
\fI\%mongerfiles\fP, you must run \fI\%mongerfiles\fP against a running
\fBmongerd\fP or \fBmongers\fP instance as appropriate.

.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBmongerfiles\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs \fBmongerfiles\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
output from database commands
.IP \(bu 2
replication activity
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmongerfiles\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-uri <connectionString>
New in version 3.4.6.

.sp
Specify a resolvable URI
connection string for the \fBmongerd\fP to which to
connect.
.sp
The following is the standard
URI connection scheme:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
mongerdb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For detailed explanations of the components of this string, refer to
the
Connection String URI Format
documentation.
.sp
\fBIMPORTANT:\fP
.INDENT 7.0
.INDENT 3.5
The following \fI\%mongerfiles\fP options are incompatible with the
\fB\-\-uri\fP option. Instead, specify these options as part of your
\fB\-\-uri\fP connection string when applicable:
.INDENT 0.0
.IP \(bu 2
\fB\-\-host\fP
.IP \(bu 2
\fB\-\-port\fP
.IP \(bu 2
\fB\-\-db\fP
.IP \(bu 2
\fB\-\-username\fP
.IP \(bu 2
\fB\-\-password\fP (when specifying the password as part of the
URI connection string)
.IP \(bu 2
\fB\-\-authenticationDatabase\fP
.IP \(bu 2
\fB\-\-authenticationMechanism\fP
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>
Specifies a resolvable hostname for the \fBmongerd\fP that holds
your GridFS system. By default \fI\%mongerfiles\fP attempts to connect
to a MongerDB process running on the localhost port number \fB27017\fP\&.
.sp
Optionally, specify a port number to connect a MongerDB instance running
on a port other than 27017.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
\fIDefault\fP: 27017
.sp
Specifies the TCP port on which the MongerDB instance listens for
client connections.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
\fIRemoved in version 3.0.\fP
.sp
Enables IPv6 support and allows \fBmongerfiles\fP to connect to the
MongerDB instance using an IPv6 network. Prior to MongerDB 3.0, you
had to specify \fI\%\-\-ipv6\fP to use IPv6. In MongerDB 3.0 and later, IPv6
is always enabled.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBmongerd\fP or \fBmongers\fP that has
TLS/SSL support enabled.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
Starting in version 3.4, if \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP is not
specified and you are not using x.509 authentication, the
system\-wide CA certificate store will be used when connecting to an
TLS/SSL\-enabled server.
.sp
If using x.509 authentication, \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP
must be specified.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
\fBVersion 3.2 and earlier:\fP For TLS/SSL connections (\fB\-\-ssl\fP) to
\fBmongerd\fP and \fBmongers\fP, if the \fBmongerfiles\fP runs without the
\fI\%\-\-sslCAFile\fP, \fBmongerfiles\fP will not attempt
to validate the server certificates. This creates a vulnerability
to expired \fBmongerd\fP and \fBmongers\fP certificates as
well as to foreign processes posing as valid \fBmongerd\fP or
\fBmongers\fP instances. Ensure that you \fIalways\fP specify the
CA file to validate the server certificates in cases where
intrusion is a possibility.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the TLS/SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\%\-\-ssl\fP option to connect
to a \fBmongerd\fP or \fBmongers\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fI\%\-\-sslPEMKeyFile\fP). Use the \fI\%\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBmongerfiles\fP will
redact the password from all logging and reporting output.
.sp
If the private key in the PEM file is encrypted and you do not specify
the \fI\%\-\-sslPEMKeyPassword\fP option, the \fBmongerfiles\fP will prompt for a passphrase. See
ssl\-certificate\-password\&.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, MongerDB logs as a
warning the use of the invalid certificate.
.sp
Starting in MongerDB 4.0, if you specify
\fB\-\-sslAllowInvalidCertificates\fP or \fBssl.allowInvalidCertificates:
true\fP when using x.509 authentication, an invalid certificate is
only sufficient to establish a TLS/SSL connection but is
\fIinsufficient\fP for authentication.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
For TLS/SSL connections to \fBmongerd\fP and
\fBmongers\fP, avoid using
\fB\-\-sslAllowInvalidCertificates\fP if possible and only use
\fB\-\-sslAllowInvalidCertificates\fP on systems where intrusion is
not possible.
.sp
If the \fBmonger\fP shell (and other
mongerdb\-tools\-support\-ssl) runs with the
\fB\-\-sslAllowInvalidCertificates\fP option, the
\fBmonger\fP shell (and other
mongerdb\-tools\-support\-ssl) will not attempt to validate
the server certificates. This creates a vulnerability to expired
\fBmongerd\fP and \fBmongers\fP certificates as
well as to foreign processes posing as valid
\fBmongerd\fP or \fBmongers\fP instances.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in TLS/SSL certificates. Allows
\fBmongerfiles\fP to connect to MongerDB instances even if the hostname in their
certificates do not match the specified hostname.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBmongerfiles\fP to use the FIPS mode of the installed OpenSSL
library. Your system must have a FIPS compliant OpenSSL library to use
the \fI\%\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS\-compatible TLS/SSL is
available only in \fI\%MongerDB Enterprise\fP\&. See
/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a MongerDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a MongerDB database
that uses authentication. Use in conjunction with the \fB\-\-username\fP and
\fB\-\-authenticationDatabase\fP options.
.sp
Changed in version 3.0.0: If you do not specify an argument for \fI\%\-\-password\fP, \fBmongerfiles\fP returns
an error.

.sp
Changed in version 3.0.2: If you wish \fBmongerfiles\fP to prompt the user
for the password, pass the \fI\%\-\-username\fP option without

debian/changelog
mongerdb (2.5.4) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/12898

 -- Ernie Hershey <ernie.hershey@mongerdb.com>  Thu, 19 Dec 2013 00:41:46 -0500

mongerdb (2.1.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10894

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jun 2012 16:56:28 -0500

mongerdb (1.9.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Thu, 11 Aug 2011 16:56:28 -0500

mongerdb (1.9.1) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jul 2011 16:56:28 -0500

mongerdb (1.9.0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10232

 -- Richard Kreuter <richard@10gen.com>  Sun, 01 May 2011 16:56:28 -0500

mongerdb (1.8.0-rc0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10245

 -- Richard Kreuter <richard@10gen.com>  Tue, 22 Feb 2011 16:56:28 -0500

mongerdb (1.7.6) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10223

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.5) unstable; urgency=low

  * durability journalingin place
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10214

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.4) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10210

 -- Richard Kreuter <richard@10gen.com>  Mon, 20 Dec 2010 16:56:28 -0500

mongerdb (1.7.3) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10195

 -- Richard Kreuter <richard@10gen.com>  Tue, 16 Nov 2010 16:56:28 -0500

mongerdb (1.7.2) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10194

 -- Richard Kreuter <richard@10gen.com>  Fri, 29 Oct 2010 16:56:28 -0500

mongerdb (1.7.1) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10193

 -- Richard Kreuter <richard@10gen.com>  Tue, 28 Sep 2010 16:56:28 -0500

mongerdb (1.7.0) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 03 Sep 2010 16:56:28 -0500

mongerdb (1.5.8) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Tue, 03 Aug 2010 16:56:28 -0500

mongerdb (1.5.7) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 30 Jul 2010 16:56:28 -0500


mongerdb (1.5.6) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10179

 -- Richard Kreuter <richard@10gen.com>  Sat, 24 Jul 2010 16:56:28 -0500

mongerdb (1.5.5) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 16 Jul 2010 16:56:28 -0500

mongerdb (1.5.4) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 2 Jul 2010 16:56:28 -0500

mongerdb (1.5.3) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Thu, 17 Jun 2010 16:56:28 -0500

mongerdb (1.5.2) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10143

 -- Richard Kreuter <richard@10gen.com>  Wed, 27 May 2010 16:56:28 -0500

mongerdb (1.5.1) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10142

 -- Richard Kreuter <richard@10gen.com>  Wed, 3 May 2010 16:56:28 -0500

mongerdb (1.5.0) unstable; urgency=low

  * replication w & real-time, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10125

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500


mongerdb (1.3.5) unstable; urgency=low

  * bug fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500

mongerdb (1.3.4) unstable; urgency=low

  * bufg fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 17 Mar 2010 16:56:28 -0500

mongerdb (1.3.3) unstable; urgency=low

  * geo

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongerdb (1.3.2) unstable; urgency=low

  * munged debian files

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongerdb (1.3.1) unstable; urgency=low

  * Initial release

 -- Kristina Chodorow <kristina@10gen.com>  Tue, 07 Apr 2009 10:18:58 -0400



src/mongo/gotools/src/github.com/mongodb/mongo-tools/mongofiles/mongofiles.go
// Copyright (C) MongoDB, Inc. 2014-present.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License. You may obtain
// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

// Package mongerfiles provides an interface to GridFS collections in a MongerDB instance.
package mongerfiles

import (
	"context"
	"fmt"
	"io"
	"os"
	"regexp"

	"github.com/mongerdb/monger-tools-common/bsonutil"
	"github.com/mongerdb/monger-tools-common/db"
	"github.com/mongerdb/monger-tools-common/json"
	"github.com/mongerdb/monger-tools-common/log"
	"github.com/mongerdb/monger-tools-common/options"
	"github.com/mongerdb/monger-tools-common/util"
	"go.mongerdb.org/monger-driver/bson"
	"go.mongerdb.org/monger-driver/bson/primitive"
	"go.mongerdb.org/monger-driver/monger/gridfs"
	driverOptions "go.mongerdb.org/monger-driver/monger/options"
)

// List of possible commands for mongerfiles.
const (
	List     = "list"
	Search   = "search"
	Put      = "put"
	PutID    = "put_id"
	Get      = "get"
	GetID    = "get_id"
	Delete   = "delete"
	DeleteID = "delete_id"
)

// MongerFiles is a container for the user-specified options and
// internal state used for running mongerfiles.
type MongerFiles struct {
	// generic monger tool options
	ToolOptions *options.ToolOptions

	// mongerfiles-specific storage options
	StorageOptions *StorageOptions

	// mongerfiles-specific input options
	InputOptions *InputOptions

	// for connecting to the db
	SessionProvider *db.SessionProvider

	// command to run
	Command string

	// filename in GridFS
	FileName string

	// ID to put into GridFS
	Id string

	// GridFS bucket to operate on
	bucket *gridfs.Bucket
}

// New constructs a new mongerfiles instance from the provided options. Will fail if cannot connect to server or if the
// provided options are invalid.
func New(opts Options) (*MongerFiles, error) {
	// create a session provider to connect to the db
	provider, err := db.NewSessionProvider(*opts.ToolOptions)
	if err != nil {
		return nil, util.SetupError{Err: fmt.Errorf("error connecting to host: %v", err)}
	}

	mf := &MongerFiles{
		ToolOptions:     opts.ToolOptions,
		StorageOptions:  opts.StorageOptions,
		SessionProvider: provider,
		InputOptions:    opts.InputOptions,
	}

	if err := mf.ValidateCommand(opts.ParsedArgs); err != nil {
		return nil, util.SetupError{Err: err, Message: util.ShortUsage("mongerfiles")}
	}

	return mf, nil
}

// Close disconnects from the server and cleans up internal mongerfiles state.
func (mf *MongerFiles) Close() {
	mf.SessionProvider.Close()
}

// ValidateCommand ensures the arguments supplied are valid.
func (mf *MongerFiles) ValidateCommand(args []string) error {
	// make sure a command is specified and that we don't have
	// too many arguments
	if len(args) == 0 {
		return fmt.Errorf("no command specified")
	}

	switch args[0] {
	case List:
		if len(args) > 2 {
			return fmt.Errorf("too many positional arguments")
		}
		if len(args) == 1 {
			mf.FileName = ""
		} else {
			mf.FileName = args[1]
		}
	case Search, Put, Get, Delete:
		if len(args) > 2 {
			return fmt.Errorf("too many positional arguments")
		}
		// also make sure the supporting argument isn't literally an
		// empty string for example, mongerfiles get ""
		if len(args) == 1 || args[1] == "" {
			return fmt.Errorf("'%v' argument missing", args[0])
		}
		mf.FileName = args[1]
	case GetID, DeleteID:
		if len(args) > 2 {
			return fmt.Errorf("too many positional arguments")
		}
		if len(args) == 1 || args[1] == "" {
			return fmt.Errorf("'%v' argument missing", args[0])
		}
		mf.Id = args[1]
	case PutID:
		if len(args) > 3 {
			return fmt.Errorf("too many positional arguments")
		}
		if len(args) < 3 || args[1] == "" || args[2] == "" {
			return fmt.Errorf("'%v' argument(s) missing", args[0])
		}
		mf.FileName = args[1]
		mf.Id = args[2]
	default:
		return fmt.Errorf("'%v' is not a valid command", args[0])
	}

	if mf.StorageOptions.GridFSPrefix == "" {
		return fmt.Errorf("--prefix can not be blank")
	}

	mf.Command = args[0]
	return nil
}

// Query GridFS for files and display the results.
func (mf *MongerFiles) findAndDisplay(query bson.M) (string, error) {
	gridFiles, err := mf.findGFSFiles(query)
	if err != nil {
		return "", fmt.Errorf("error retrieving list of GridFS files: %v", err)
	}

	var display string
	for _, gridFile := range gridFiles {
		display += fmt.Sprintf("%s\t%d\n", gridFile.Name, gridFile.Length)
	}

	return display, nil
}

// Return the local filename, as specified by the --local flag. Defaults to
// the GridFile's name if not present. If GridFile is nil, uses the filename
// given on the command line.
func (mf *MongerFiles) getLocalFileName(gridFile *gfsFile) string {
	localFileName := mf.StorageOptions.LocalFileName
	if localFileName == "" {
		if gridFile != nil {
			localFileName = gridFile.Name
		} else {
			localFileName = mf.FileName
		}
	}
	return localFileName
}

// handleGet contains the logic for the 'get' and 'get_id' commands
func (mf *MongerFiles) handleGet() (err error) {
	file, err := mf.getTargetGFSFile()
	if err != nil {
		return err
	}

	if err = mf.writeGFSFileToLocal(file); err != nil {
		return err
	}

	return err
}

// Gets all GridFS files that match the given query.
func (mf *MongerFiles) findGFSFiles(query bson.M) (files []*gfsFile, err error) {
	cursor, err := mf.bucket.Find(query)
	if err != nil {
		return nil, err
	}
	dc := util.DeferredCloser{Closer: &util.CloserCursor{Cursor: cursor}}
	defer dc.CloseWithErrorCapture(&err)

	for cursor.Next(context.Background()) {
		var out *gfsFile
		out, err = newGfsFileFromCursor(cursor, mf)
		if err != nil {
			return nil, err
		}
		files = append(files, out)
	}

	return files, nil
}

// Gets the GridFS file the options specify. Use this for the get family of commands.
func (mf *MongerFiles) getTargetGFSFile() (*gfsFile, error) {
	var gridFiles []*gfsFile
	var err error

	var queryProp string
	var query string

	if mf.Id != "" {
		queryProp = "_id"
		query = mf.Id

		id, err := mf.parseOrCreateID()
		if err != nil {
			return nil, err
		}
		gridFiles, err = mf.findGFSFiles(bson.M{"_id": id})
		if err != nil {
			return nil, err
		}
	} else {
		queryProp = "name"
		query = mf.FileName

		gridFiles, err = mf.findGFSFiles(bson.M{"filename": mf.FileName})
		if err != nil {
			return nil, err
		}
	}

	if len(gridFiles) == 0 {
		return nil, fmt.Errorf("no such file with %v: %v", queryProp, query)
	}

	return gridFiles[0], nil
}

// Delete all files with the given filename.
func (mf *MongerFiles) deleteAll(filename string) error {
	gridFiles, err := mf.findGFSFiles(bson.M{"filename": filename})
	if err != nil {
		return err
	}

	for _, gridFile := range gridFiles {
		if err := gridFile.Delete(); err != nil {
			return err
		}
	}
	log.Logvf(log.Always, "successfully deleted all instances of '%v' from GridFS\n", mf.FileName)

	return nil
}

// handleDeleteID contains the logic for the 'delete_id' command
func (mf *MongerFiles) handleDeleteID() error {
	file, err := mf.getTargetGFSFile()
	if err != nil {
		return err
	}

	if err := file.Delete(); err != nil {
		return err
	}
	log.Logvf(log.Always, fmt.Sprintf("successfully deleted file with _id %v from GridFS", mf.Id))

	return nil
}

// parse and convert input extended JSON _id. Generates a new ObjectID if no _id provided.
func (mf *MongerFiles) parseOrCreateID() (interface{}, error) {
	if mf.Id == "" {
		return primitive.NewObjectID(), nil
	}

	var asJSON interface{}
	if err := json.Unmarshal([]byte(mf.Id), &asJSON); err != nil {
		return nil, fmt.Errorf("error parsing provided extJSON: %v", err)
	}

	// legacy extJSON parser
	id, err := bsonutil.ConvertLegacyExtJSONValueToBSON(asJSON)
	if err != nil {
		return nil, fmt.Errorf("error converting extJSON vlaue to bson: %v", err)
	}

	return id, nil
}

// writeGFSFileToLocal writes a file from gridFS to stdout or the filesystem.
func (mf *MongerFiles) writeGFSFileToLocal(gridFile *gfsFile) (err error) {
	localFileName := mf.getLocalFileName(gridFile)
	var localFile io.WriteCloser
	if localFileName == "-" {
		localFile = os.Stdout
	} else {
		if localFile, err = os.Create(localFileName); err != nil {
			return fmt.Errorf("error while opening local file '%v': %v", localFileName, err)
		}
		dc := util.DeferredCloser{Closer: localFile}
		defer dc.CloseWithErrorCapture(&err)
		log.Logvf(log.DebugLow, "created local file '%v'", localFileName)
	}

	stream, err := gridFile.OpenStreamForReading()
	if err != nil {
		return err
	}
	dc := util.DeferredCloser{Closer: stream}
	defer dc.CloseWithErrorCapture(&err)

	if _, err = io.Copy(localFile, stream); err != nil {
		return fmt.Errorf("error while writing Data into local file '%v': %v", localFileName, err)
	}

	log.Logvf(log.Always, fmt.Sprintf("finished writing to %s\n", localFileName))
	return nil
}

// Write the given GridFS file to the database. Will fail if file already exists and --replace flag turned off.
func (mf *MongerFiles) put(id interface{}, name string) (bytesWritten int64, err error) {
	gridFile, err := newGfsFile(id, name, mf)
	if err != nil {
		return 0, err
	}

	localFileName := mf.getLocalFileName(gridFile)

	var localFile io.ReadCloser
	if localFileName == "-" {
		localFile = os.Stdin
	} else {
		localFile, err = os.Open(localFileName)
		if err != nil {
			return 0, fmt.Errorf("error while opening local gridFile '%v' : %v", localFileName, err)
		}
		dc := util.DeferredCloser{Closer: localFile}
		defer dc.CloseWithErrorCapture(&err)
		log.Logvf(log.DebugLow, "creating GridFS gridFile '%v' from local gridFile '%v'", mf.FileName, localFileName)
	}

	// check if --replace flag turned on
	if mf.StorageOptions.Replace {
		if err = mf.deleteAll(gridFile.Name); err != nil {
			return 0, err
		}
	}

	if mf.StorageOptions.ContentType != "" {
		gridFile.Metadata.ContentType = mf.StorageOptions.ContentType
	}

	stream, err := gridFile.OpenStreamForWriting()
	if err != nil {
		return 0, err
	}
	dc := util.DeferredCloser{Closer: stream}
	defer dc.CloseWithErrorCapture(&err)

	n, err := io.Copy(stream, localFile)
	if err != nil {
		return n, fmt.Errorf("error while storing '%v' into GridFS: %v", localFileName, err)
	}

	return n, nil
}

// handlePut contains the logic for the 'put' and 'put_id' commands
func (mf *MongerFiles) handlePut() error {
	id, err := mf.parseOrCreateID()
	if err != nil {
		return err
	}

	n, err := mf.put(id, mf.FileName)
	if err != nil {
		return err
	}

	log.Logvf(log.DebugLow, "copied %v bytes to server", n)
	log.Logvf(log.Always, fmt.Sprintf("added gridFile: %v\n", mf.FileName))

	return nil
}

// Run the mongerfiles utility. If displayHost is true, the connected host/port is
// displayed.
func (mf *MongerFiles) Run(displayHost bool) (output string, finalErr error) {
	var err error

	// check type of node we're connected to, and fall back to w=1 if standalone (for <= 2.4)
	nodeType, err := mf.SessionProvider.GetNodeType()
	if err != nil {
		return "", fmt.Errorf("error determining type of node connected: %v", err)
	}

	log.Logvf(log.DebugLow, "connected to node type: %v", nodeType)

	client, err := mf.SessionProvider.GetSession()
	if err != nil {
		return "", fmt.Errorf("error getting client: %v", err)
	}

	err = client.Ping(context.Background(), nil)
	if err != nil {
		return "", fmt.Errorf("error connecting to host: %v", err)
	}

	database := client.Database(mf.StorageOptions.DB)
	mf.bucket, err = gridfs.NewBucket(database, &driverOptions.BucketOptions{Name: &mf.StorageOptions.GridFSPrefix})
	if err != nil {
		return "", fmt.Errorf("error getting GridFS bucket: %v", err)
	}

	if displayHost {
		log.Logvf(log.Always, "connected to: %v", mf.ToolOptions.URI.ConnectionString)
	}

	// first validate the namespaces we'll be using: <db>.<prefix>.files and <db>.<prefix>.chunks
	// it's ok to validate only <db>.<prefix>.chunks (the longer one)
	err = util.ValidateFullNamespace(fmt.Sprintf("%s.%s.chunks", mf.StorageOptions.DB,
		mf.StorageOptions.GridFSPrefix))
	if err != nil {
		return "", err
	}

	log.Logvf(log.Info, "handling mongerfiles '%v' command...", mf.Command)

	switch mf.Command {

	case List:
		query := bson.M{}
		if mf.FileName != "" {
			regex := bson.M{"$regex": "^" + regexp.QuoteMeta(mf.FileName)}
			query = bson.M{"filename": regex}
		}
		output, err = mf.findAndDisplay(query)

	case Search:
		regex := bson.M{"$regex": mf.FileName}
		query := bson.M{"filename": regex}

		output, err = mf.findAndDisplay(query)

	case Get, GetID:
		err = mf.handleGet()

	case Put, PutID:
		err = mf.handlePut()

	case DeleteID:
		err = mf.handleDeleteID()

	case Delete:
		err = mf.deleteAll(mf.FileName)
	}

	return output, err
}
